server/service/src/invoice_line/save_outbound_shipment_item_lines/generate.rs

```diff
@@ -0,0 +1,153 @@
+                batch: None,
+                pack_size: None,
+                expiry_date: None,
+                prescribed_quantity: None,
+                total_before_tax: None,
+                tax_percentage: None,
+                note: None,
+                location_id: None,
+                cost_price_per_pack: None,
+                sell_price_per_pack: None,
+            },
+        )
+        .collect();
+
+    let lines_to_update = lines
+        .clone()
+        .into_iter()
+        .filter(|line| line.number_of_packs > 0.0 && check_already_exists(&line.id))
+        .map(
+            |SaveOutboundShipmentLine {
+                 id,
+                 number_of_packs,
+                 stock_line_id,
+             }| UpdateStockOutLine {
+                id,
+                stock_line_id: Some(stock_line_id),
+                number_of_packs: Some(number_of_packs),
+                r#type: Some(StockOutType::OutboundShipment),
+                // Default
+                prescribed_quantity: None,
+                total_before_tax: None,
+                tax: None,
+                note: None,
+            },
+        )
+        .collect();
+
+    let lines_to_delete = lines
+        .clone()
+        .into_iter()
+        .filter(|line| line.number_of_packs <= 0.0 && check_already_exists(&line.id))
+        .map(|line| DeleteStockOutLine {
+            id: line.id,
+            r#type: Some(StockOutType::OutboundShipment),
+        })
+        .collect();
+
+    let existing_placeholder = existing_lines
+        .iter()
+        .find(|line| line.invoice_line_row.r#type == InvoiceLineType::UnallocatedStock);
+
+    let manage_placeholder = match (existing_placeholder, placeholder_quantity) {
+        (Some(line), Some(quantity)) => {
+            if quantity > 0.0 {
+                ManagePlaceholderLine::Update(UpdateOutboundShipmentUnallocatedLine {
+                    id: line.invoice_line_row.id.clone(),
+                    quantity,
+                })
+            } else {
+                ManagePlaceholderLine::Delete(DeleteOutboundShipmentUnallocatedLine {
+                    id: line.invoice_line_row.id.clone(),
+                })
+            }
+        }
+        (Some(line), None) => {
+            ManagePlaceholderLine::Delete(DeleteOutboundShipmentUnallocatedLine {
+                id: line.invoice_line_row.id.clone(),
+            })
+        }
+        (None, Some(quantity)) if quantity > 0.0 => {
+            ManagePlaceholderLine::Insert(InsertOutboundShipmentUnallocatedLine {
+                id: uuid::uuid(),
+                invoice_id,
+                item_id,
+                quantity: quantity as u32, // worth cleaning up this input... gets saved as f64?
+            })
+        }
+        _ => ManagePlaceholderLine::NothingToDo,
+    };
+
+    Ok(GenerateResult {
+        lines_to_add,
+        lines_to_update,
+        lines_to_delete,
```

Comment:
I think we would still need to check (I think it's ok in carry over or cool down):
* Duplicated stock lines
* Stock lines in store
* Stock line belongs to that item
* Trip all invoice_lines that are not in the payload (for that item)